### PR TITLE
BUG: Fix `_extent` not set in PcolorImage

### LINKED ISSUE #8426: PcolorImage does not set `_extent`
To help us understand and resolve your issue, please fill out the form to the best of your ability. You can feel free to delete the sections that do not apply.

### Bug report

**Bug summary**

Creating a simple 2d histogram fails when the bins are not equally spaced under some circumstances.

**Code for reproduction**

- A minimum code snippet required to reproduce the bug, also minimizing the number of dependencies required

On a Jupyter notebook

```python
%matplotlib inline
import matplotlib.pyplot as plt
x = plt.hist2d([1,2,3], [3,5,6], bins=[[0,3,7], [1,2,3]])
```

**Actual outcome**

- The output produced by the above code, which may be a screenshot, console output, etc.

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/home/zah/anaconda3/lib/python3.6/site-packages/IPython/core/formatters.py in __call__(self, obj)
    305                 pass
    306             else:
--> 307                 return printer(obj)
    308             # Finally look for special method names
    309             method = get_real_method(obj, self.print_method)

/home/zah/anaconda3/lib/python3.6/site-packages/IPython/core/pylabtools.py in <lambda>(fig)
    238 
    239     if 'png' in formats:
--> 240         png_formatter.for_type(Figure, lambda fig: print_figure(fig, 'png', **kwargs))
    241     if 'retina' in formats or 'png2x' in formats:
    242         png_formatter.for_type(Figure, lambda fig: retina_figure(fig, **kwargs))

/home/zah/anaconda3/lib/python3.6/site-packages/IPython/core/pylabtools.py in print_figure(fig, fmt, bbox_inches, **kwargs)
    122 
    123     bytes_io = BytesIO()
--> 124     fig.canvas.print_figure(bytes_io, **kw)
    125     data = bytes_io.getvalue()
    126     if fmt == 'svg':

/home/zah/anaconda3/lib/python3.6/site-packages/matplotlib/backend_bases.py in print_figure(self, filename, dpi, facecolor, edgecolor, orientation, format, **kwargs)
   2200                 bbox_filtered = []
   2201                 for a in bbox_artists:
-> 2202                     bbox = a.get_window_extent(renderer)
   2203                     if a.get_clip_on():
   2204                         clip_box = a.get_clip_box()

/home/zah/anaconda3/lib/python3.6/site-packages/matplotlib/image.py in get_window_extent(self, renderer)
    704 
    705     def get_window_extent(self, renderer=None):
--> 706         x0, x1, y0, y1 = self._extent
    707         bbox = Bbox.from_extents([x0, y0, x1, y1])
    708         return bbox.transformed(self.axes.transData)

TypeError: 'NoneType' object is not iterable

<matplotlib.figure.Figure at 0x7f828c153240>
```
This works if I remove the `%matplotlib inline` above, even if I try to write a png image with e.g. `plt.savefig('/tmp/xx.png')`. However the bug was found in a code that doesn't use IPython (though I don't know how to reproduce it).

It does work if I change the histogram to:

```python
x = plt.hist2d([1,2,3], [3,5,6], bins=[[0,3,6], [1,2,3]])
```

making all the bins equally spaced.

**Expected outcome**

- No crash.

I'm using the conda forge package of mpl 2.0.

### PR DESCRIPTION
**Pull request by AtharvaKhare**
PcolorImage now sets `_extent` of AxesImage on its creation.

## PR Summary

Depending on bins argument of hist2d, style is set to either `image` or `pcolorimage` in _axes.pcolorfast. If the style is `image`, AxesImage object is created and is supplied with extents. The issue arose when style was `pcolorimage`. The `_extent` is never set in super class AxesImage of PcolorImage. 
The issue is solved by setting `_extent` after PcolorImage creation.

This PR fixes #8426.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is PEP 8 compliant 
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by jklymak** [comment]
Thanks for this!  Can you supply a test that doesn't work w/ master but does work with your suggested change?  It'd help (me?) to get context for it.

**Comment by AtharvaKhare** [comment]
@jklymak Sample example from issue:
```
import pylab as plt
x = plt.hist2d([1,2,3], [3,5,6], bins=[[0,3,7], [1,2,3]])
im = x[-1]
print(type(im))
print(im.get_window_extent())
```
The issue arose when plt.hist2d returns object of `matplotlib.image.PcolorImage`. 
hist2d returns either `matplotlib.image.AxesImage` or `matplotlib.image.PcolorImage` depending on the parameter bins. The condition for selection is:
https://github.com/matplotlib/matplotlib/blob/d7ddd479c2fb6c10fd4482d884f783fc1458c6ac/lib/matplotlib/axes/_axes.py#L5781-L5795
Followed by creation of object:
https://github.com/matplotlib/matplotlib/blob/d7ddd479c2fb6c10fd4482d884f783fc1458c6ac/lib/matplotlib/axes/_axes.py#L5837-L5842
or
https://github.com/matplotlib/matplotlib/blob/d7ddd479c2fb6c10fd4482d884f783fc1458c6ac/lib/matplotlib/axes/_axes.py#L5845-L5850
Here extent was not specified while creation of the PcolorImage object.
https://github.com/matplotlib/matplotlib/blob/d7ddd479c2fb6c10fd4482d884f783fc1458c6ac/lib/matplotlib/image.py#L980
PcolorImage calls AxesImage constructor with extent=None, which causes error when referenced.

In short, to recreate error, create any PcolorImage object and reference its parent's(AxesImage) extents. To do this, elements of bins must be of dim 1 and size >=3, with intervals such that the do not satisfy: 
https://github.com/matplotlib/matplotlib/blob/d7ddd479c2fb6c10fd4482d884f783fc1458c6ac/lib/matplotlib/axes/_axes.py#L5789-L5795
Basically some non-equally spaced bins.

**Comment by dstansby** [comment]
Could you add the test in `lib/matplotlib/tests/test_image.py`? It doesn't need to be an image test, but just a test to check that the code doesn't fail.

**Comment by AtharvaKhare** [comment]
Changed the way `_extent` is set.

@dstansby Added test.
